qiskit/circuit/library/diagonal.py

```diff
@@ -0,0 +1,124 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# The structure of the code is based on Emanuel Malvetti's semester thesis at ETH in 2018,
+# which was supervised by Raban Iten and Prof. Renato Renner.
+
+
+"""Diagonal matrix circuit."""
+
+from typing import Union, List, Optional
+import cmath
+import numpy as np
+
+from qiskit.circuit.quantumcircuit import QuantumCircuit
+from qiskit.circuit.exceptions import CircuitError
+
+_EPS = 1e-10
+
+
+class Diagonal(QuantumCircuit):
+    r"""Diagonal circuit.
+
+    Circuit symbol:
+
+    .. parsed-literal::
+
+             ┌───────────┐
+        q_0: ┤0          ├
+             │           │
+        q_1: ┤1 Diagonal ├
+             │           │
+        q_2: ┤2          ├
+             └───────────┘
+
+    Matrix form:
+
+    .. math::
+        DiagonalGate\ q_0, q_1, .., q_{n-1} =
```

Comment:
Does this render nicely or will the test be in cursive math font? Otherwise we could put it in `text` or `mathrm`:
```suggestion
        \text{DiagonalGate}\ q_0, q_1, .., q_{n-1} =
```